src/cmap/connection.ts

```diff
@@ -1171,26 +1009,130 @@ export class ModernConnection extends TypedEventEmitter<ConnectionEvents> {
 
-    const commandOptions: Document = Object.assign(
-      {
-        numberToSkip: 0,
-        numberToReturn: -1,
-        checkKeys: false,
-        // This value is not overridable
-        secondaryOk: readPreference.secondaryOk()
-      },
-      options
-    );
+    const commandOptions = {
+      numberToSkip: 0,
+      numberToReturn: -1,
+      checkKeys: false,
+      // This value is not overridable
+      secondaryOk: readPreference.secondaryOk(),
+      ...options,
+      readPreference // ensure we pass in ReadPreference instance
+    };
 
     const message = this.supportsOpMsg
-      ? new OpMsgRequest(ns.db, cmd, commandOptions)
-      : new OpQueryRequest(ns.db, cmd, commandOptions);
+      ? new OpMsgRequest(db, cmd, commandOptions)
+      : new OpQueryRequest(db, cmd, commandOptions);
+
+    return message;
+  }
+
+  private async sendCommand(
+    message: WriteProtocolMessageType,
+    options: CommandOptions
+  ): Promise<Document> {
+    const { signal } = this.controller;
 
+    signal.throwIfAborted();
+
+    if (typeof options.socketTimeoutMS === 'number') {
+      this.socket.setTimeout(options.socketTimeoutMS);
+    } else if (this.socketTimeoutMS !== 0) {
+      this.socket.setTimeout(this.socketTimeoutMS);
+    }
+
+    let response;
     try {
-      write(this as any as Connection, message, commandOptions, callback);
-    } catch (err) {
-      callback(err);
+      await writeCommand(this, message, {
+        agreedCompressor: this.description.compressor ?? 'none',
+        zlibCompressionLevel: this.description.zlibCompressionLevel,
+        signal
+      });
+
+      if (options.noResponse) return { ok: 1 };
+
+      response = await read(this, { signal });
+    } finally {
+      if (!signal.aborted) this.controller = new AbortController();
     }
+
+    response.parse(options);
+
+    const [document] = response.documents;
+
+    if (!Buffer.isBuffer(document)) {
+      const { session } = options;
+      if (session) {
+        updateSessionFromResponse(session, document);
+      }
+
+      if (document.$clusterTime) {
+        this[kClusterTime] = document.$clusterTime;
+        this.emit(Connection.CLUSTER_TIME_RECEIVED, document.$clusterTime);
+      }
+    }
+
+    return document;
+  }
+
+  async command(
+    ns: MongoDBNamespace,
+    command: Document,
+    options: CommandOptions = {}
+  ): Promise<Document> {
+    // Temporary to make sure no callback usage:
+    // eslint-disable-next-line prefer-rest-params
+    if (typeof arguments[arguments.length - 1] === 'function') {
+      throw new Error('no callbacks allowed!!');
+    }
+
+    const message = this.prepareCommand(ns.db, command, options);
+
+    let started = 0;
+    if (this.monitorCommands) {
+      started = now();
+      this.emit(
+        ModernConnection.COMMAND_STARTED,
+        new CommandStartedEvent(this as unknown as Connection, message)
+      );
+    }
+
+    const document = await this.sendCommand(message, options);
```

Comment:
What happens with command monitoring if `sendCommand` rejects?  I think we need a `catch` block that records a command failed event